File: shuffler/src/util/encrypted_message_util_test.go
// Copyright 2017 The Fuchsia Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package util

import (
	"reflect"
	"testing"

	"cobalt"
)

var privateKeyPem, publicKeyPem string

func init() {
	privateKeyPem = `-----BEGIN PRIVATE KEY-----
MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQg1kZxvT81qrRWg2Y8
g/M7YNtiHaC14/fbevhy/hgXcByhRANCAASkbLO+7iLLaPayYIr3YVmY0jkbwalG
sOB9Tf3R8TR7Ow43cHlGjX3HALV1z4Lxs1v2K13yeegBJF8lU88cdAqY
-----END PRIVATE KEY-----`

	publicKeyPem = `-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEpGyzvu4iy2j2smCK92FZmNI5G8Gp
RrDgfU390fE0ezsON3B5Ro19xwC1dc+C8bNb9itd8nnoASRfJVPPHHQKmA==
-----END PUBLIC KEY-----`
}

// Makes and returns an Envelope with some non-default values so that we can recognized
// when we correctly encrypt and decrypt it.
func MakeTestEnvelope() cobalt.Envelope {
	return cobalt.Envelope{
		Batch: []*cobalt.ObservationBatch{
			&cobalt.ObservationBatch{
				MetaData: &cobalt.ObservationMetadata{
					CustomerId: 42,
					ProjectId:  43,
					MetricId:   44,
				},
			},
		},
	}
}

// Tests that a MessageDecrypter that is constructed without a valid private
// key can decrypt messages that use the NONE scheme.
func TestNoEncryption(t *testing.T) {
	// Make an EncryptedMessageMaker
	encryptedMessageMaker := NewEncryptedMessageMaker("", cobalt.EncryptedMessage_NONE)
	if encryptedMessageMaker == nil {
		t.Fatal("Failed to create EncryptedMessageMaker")
	}

	// Make an Envelope with some non-default values so we can recognize it.
	envelope1 := MakeTestEnvelope()

	// Encrypt the Envelope
	encryptedMessage, err := encryptedMessageMaker.Encrypt(&envelope1)
	if err != nil {
		t.Errorf("%v", err)
	}

	// Make a MessageDecrypter
	messageDecrypter := NewMessageDecrypter("")

	// Decrypt the Envelope
	envelope2 := cobalt.Envelope{}
	err = messageDecrypter.DecryptMessage(encryptedMessage, &envelope2)
	if err != nil {
		t.Errorf("%v", err)
	}

	// Compare the recovered envelope with the plaintext envelope
	if !reflect.DeepEqual(&envelope1, &envelope2) {
		t.Errorf("%v != %v", envelope1, envelope2)
	}
}

// Tests that a MessageDecrypter that is constructed with a valid private key
// can decrypt messages that use the HYBRID_ECDH_V1 scheme.
func TestHybridEncryption(t *testing.T) {
	// Make an EncryptedMessageMaker
	encryptedMessageMaker := NewEncryptedMessageMaker(publicKeyPem, cobalt.EncryptedMessage_HYBRID_ECDH_V1)
	if encryptedMessageMaker == nil {
		t.Fatal("Failed to create EncryptedMessageMaker")
	}

	// Make an Envelope with some non-default values so we can recognize it.
	envelope1 := MakeTestEnvelope()

	// Encrypt the Envelope
	encryptedMessage, err := encryptedMessageMaker.Encrypt(&envelope1)
	if err != nil {
		t.Errorf("%v", err)
	}

	// Make a MessageDecrypter
	messageDecrypter := NewMessageDecrypter(privateKeyPem)

	// Decrypt the Envelope
	envelope2 := cobalt.Envelope{}
	err = messageDecrypter.DecryptMessage(encryptedMessage, &envelope2)
	if err != nil {
		t.Errorf("%v", err)
	}

	// Compare the recovered envelope with the plaintext envelope
	if !reflect.DeepEqual(&envelope1, &envelope2) {
		t.Errorf("%v != %v", envelope1, envelope2)
	}
}

// Tests that a MessageDecrypter that is constructed with an invalid private key
// can fail gracefully.
func TestFailedHybridEncryption(t *testing.T) {
	// Make an EncryptedMessageMaker
	encryptedMessageMaker := NewEncryptedMessageMaker(publicKeyPem, cobalt.EncryptedMessage_HYBRID_ECDH_V1)
	if encryptedMessageMaker == nil {
		t.Fatal("Failed to create EncryptedMessageMaker")
	}

	// Make an Envelope with some non-default values so we can recognize it.
	envelope1 := MakeTestEnvelope()

	// Encrypt the Envelope
	encryptedMessage, err := encryptedMessageMaker.Encrypt(&envelope1)
	if err != nil {
		t.Errorf("%v", err)
	}

	// Make a MessageDecrypter, but do not give it a valid privatekeyPem.
	messageDecrypter := NewMessageDecrypter("")

	// Try to decrypt the Envelope
	envelope2 := cobalt.Envelope{}
	err = messageDecrypter.DecryptMessage(encryptedMessage, &envelope2)
	if err == nil {
		t.Errorf("Expected an error.")
	}
}

// Tests that a MessageDecrypter that is constructed with a valid private key
// and is given a corrupted ciphertext can fail gracefully.
func TestCorruptedHybridEncryption(t *testing.T) {
	// Make an EncryptedMessageMaker
	encryptedMessageMaker := NewEncryptedMessageMaker(publicKeyPem, cobalt.EncryptedMessage_HYBRID_ECDH_V1)
	if encryptedMessageMaker == nil {
		t.Fatal("Failed to create EncryptedMessageMaker")
	}

	// Make an Envelope with some non-default values so we can recognize it.
	envelope1 := MakeTestEnvelope()

	// Encrypt the Envelope
	encryptedMessage, err := encryptedMessageMaker.Encrypt(&envelope1)
	if err != nil {
		t.Errorf("%v", err)
	}

	// Make a MessageDecrypter
	messageDecrypter := NewMessageDecrypter(privateKeyPem)

	// Corrupt the EncrptedMessage
	encryptedMessage.Ciphertext[0] = encryptedMessage.Ciphertext[0] + 1

	// Attempt to decrypt the Envelope
	envelope2 := cobalt.Envelope{}
	err = messageDecrypter.DecryptMessage(encryptedMessage, &envelope2)

	// Check that an error was returned, but there was no crash.
	if err == nil {
		t.Errorf("Expected an error.")
	}
}

File: util/encrypted_message_util_test.cc
// Copyright 2017 The Fuchsia Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "util/encrypted_message_util.h"

#include <string>

#include "./encrypted_message.pb.h"
#include "./observation.pb.h"
#include "third_party/googletest/googletest/include/gtest/gtest.h"
#include "util/crypto_util/cipher.h"

namespace cobalt {
namespace util {

using crypto::HybridCipher;

Observation MakeDummyObservation(std::string part_name) {
  Observation observation;
  (*observation.mutable_parts())[part_name] = ObservationPart();
  return observation;
}

// Tests the use of the no-encryption option.
TEST(EncryptedMessageUtilTest, NoEncryption) {
  // Make a dummy observation.
  auto observation = MakeDummyObservation("hello");
  // Make an EncryptedMessageMaker that uses the NONE encryption scheme.
  EncryptedMessageMaker maker("dummy_key", EncryptedMessage::NONE);
  // Encrypt the dummy observation.
  EncryptedMessage encrypted_message;
  EXPECT_TRUE(maker.Encrypt(observation, &encrypted_message));

  // Make a MessageDecrypter.
  MessageDecrypter decrypter("dummy_key");
  // Decrypt and check.
  observation.Clear();
  EXPECT_TRUE(decrypter.DecryptMessage(encrypted_message, &observation));
  EXPECT_EQ(1u, observation.parts().count("hello"));
}

// Tests the use of bad encryption keys
TEST(EncryptedMessageUtilTest, BadKeys) {
  // Make a dummy observation.
  auto observation = MakeDummyObservation("hello");

  // Make an EncryptedMessageMaker that uses a bad public key.
  EncryptedMessageMaker maker("dummy_key", EncryptedMessage::HYBRID_ECDH_V1);
  // Try to encrypt the dummy observation.
  EncryptedMessage encrypted_message;
  // Expect it to fail, but not crash.
  EXPECT_FALSE(maker.Encrypt(observation, &encrypted_message));
}

// Tests the use of the hybrid cipher option.
TEST(EncryptedMessageUtilTest, HybridEncryption) {
  std::string public_key;
  std::string private_key;
  EXPECT_TRUE(HybridCipher::GenerateKeyPairPEM(&public_key, &private_key));

  // Make a dummy observation.
  auto observation = MakeDummyObservation("hello");

  // Make an EncryptedMessageMaker that uses our real encryption scheme.
  EncryptedMessageMaker maker(public_key, EncryptedMessage::HYBRID_ECDH_V1);
  // Encrypt the dummy observation.
  EncryptedMessage encrypted_message;
  ASSERT_TRUE(maker.Encrypt(observation, &encrypted_message));
  EXPECT_EQ(32u, encrypted_message.public_key_fingerprint().size());

  // Make a MessageDecrypter.
  MessageDecrypter decrypter(private_key);
  // Decrypt and check.
  observation.Clear();
  ASSERT_TRUE(decrypter.DecryptMessage(encrypted_message, &observation));
  EXPECT_EQ(1u, observation.parts().count("hello"));

  // Make a MessageDecrypter that uses a bad private key.
  MessageDecrypter bad_decrypter("dummy_key");
  // Try to decrypt.
  // Expect it to fail, but not crash.
  EXPECT_FALSE(bad_decrypter.DecryptMessage(encrypted_message, &observation));
}

// Tests that using encryption incorrectly fails but doesn't cause any crashes.
TEST(EncryptedMessageUtilTest, Crazy) {
  std::string public_key;
  std::string private_key;
  EXPECT_TRUE(HybridCipher::GenerateKeyPairPEM(&public_key, &private_key));

  // Make a dummy observation.
  auto observation = MakeDummyObservation("hello");

  // Make an EncryptedMessageMaker that incorrectly uses the private key
  // instead of the public key
  EncryptedMessageMaker bad_maker(private_key,
                                  EncryptedMessage::HYBRID_ECDH_V1);
  // Try to encrypt the dummy observation.
  EncryptedMessage encrypted_message;
  // Expect it to fail, but not crash.
  EXPECT_FALSE(bad_maker.Encrypt(observation, &encrypted_message));

  // Now make a good EncryptedMessageMaker
  EncryptedMessageMaker real_maker(public_key,
                                   EncryptedMessage::HYBRID_ECDH_V1);
  // Encrypt the dummy observation.
  EXPECT_TRUE(real_maker.Encrypt(observation, &encrypted_message));

  // Make a MessageDecrypter that uses the correct private key.
  MessageDecrypter real_decrypter(private_key);
  // Decrypt and check.
  observation.Clear();
  EXPECT_TRUE(real_decrypter.DecryptMessage(encrypted_message, &observation));
  EXPECT_EQ(1u, observation.parts().count("hello"));

  // Make a MessageDecrypter that incorrectly uses the public key
  MessageDecrypter bad_decrypter(public_key);
  // Try to decrypt.
  // Expect it to fail, but not crash.
  EXPECT_FALSE(bad_decrypter.DecryptMessage(encrypted_message, &observation));

  // Try to decrypt corrupted ciphertext.
  // Expect it to fail, but not crash.
  encrypted_message.mutable_ciphertext()[0] =
      encrypted_message.ciphertext()[0] + 1;
  EXPECT_FALSE(real_decrypter.DecryptMessage(encrypted_message, &observation));
}

}  // namespace util
}  // namespace cobalt
